system/src/system_tracer.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2017 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "concurrent_hal.h"
+#include "core_hal.h"
+#include <stdio.h>
+#include "tracer_service.h"
+
+#if defined(PLATFORM_TRACER_ENABLED)
+#include "platform_headers.h"
+#include "system_tracer_service_impl.h"
+
+#ifdef TRACER_LOCATION_DEFINITION
+TRACER_LOCATION_DEFINITION()
+#endif
+
+using namespace particle::tracer;
+
+struct TracerServiceInitializer {
+    TracerServiceInitializer() {
+        auto s = TracerService::lock(true);
+
+        auto tracer = TracerService::instance();
+        (void)tracer;
+
+        tracer_callbacks_t cb = {0};
+        cb.tracer_save_checkpoint = &tracer_save_checkpoint;
+        tracer_set_callbacks(&cb, nullptr);
+#if defined(MODULE_HAS_SYSTEM_PART3) && MODULE_HAS_SYSTEM_PART3 == 1
+        tracer_set_callbacks_(&cb, nullptr);
+#endif /* defined(MODULE_HAS_SYSTEM_PART3) && MODULE_HAS_SYSTEM_PART3 == 1 */
+
+        TracerService::unlock(true, s);
+    }
+};
+
+static TracerServiceInitializer s_initializer;
+
+struct tracer_save_data_t {
+    TracerService* tracer;
+    tracer_checkpoint_t* chkpt;
+    os_unique_id_t id;
+    uint32_t flags;
+    tracer_error_t err;
+};
+
+int tracer_save_checkpoint(tracer_checkpoint_t* chkpt, uint32_t flags, void* reserved)
+{
+    tracer_error_t err = TRACER_ERROR_NONE;
+    os_thread_t thread = os_thread_current();
+    os_unique_id_t id = os_thread_unique_id(thread);
+
+    // bool isr = HAL_IsISR();
+    bool isr = true;
```

Comment:
:+1: We are always disabling interrupts, instead of disabling os scheduler, because `tracer_save_checkpoint` will be used at the very least from hardfault and WWDG ISRs.